### PR TITLE
RISC-V: Porting riscv-p-extension.

### PR DESCRIPTION
**Pull request by zhixiao-zhang**
Main changes compared to last porting:

- riscv-builtin-rvp.def: add a suffix "_p" to each instrument to avoid the collision with riscv v extension (the patterns are defined in autovec.md and vector.md).
- rvp.md: the mov operations are defined in vector.md along with the v extension's VLS operation to avoid duplication.
- riscv.opt: now the mask flags are defined here instead of riscv-opts.h.

Extra difference with original 14.1.0 version compared to last porting:

- riscv.cc: diable a dfa reservation assert which will make the compilation failed.

Result of regression test:

Before porting:
```
                === gcc Summary ===

# of expected passes            161972
# of unexpected failures        22754
# of unexpected successes       4
# of expected failures          1435
# of unresolved testcases       20896
# of unsupported tests          13426
{path}/original/.build/gcc/xgcc  version 14.1.0 (GCC)
```

After porting:
```
                === gcc Summary ===

# of expected passes            162962
# of unexpected failures        22754
# of unexpected successes       4
# of expected failures          1435
# of unresolved testcases       20896
# of unsupported tests          13426
{path}/riscv-gcc/.build/gcc/xgcc  version 14.1.0 (GCC)
```

Build and test:

```
# prepare newlib and gcc
# Download latest newlib from ftp://sourceware.org/pub/newlib and extract it

git clone   https://github.com/forceofsystem/riscv-gcc.git
git checkout 14.1.0

cd riscv-gcc

ln -s {path to latest newlib}/newlib ./

mkdir .build && cd .build

# build gcc with p-ext

../configure --prefix=`pwd`/../../.install --target=riscv64-unknown-elf \
--disable-nls \
--disable-ssp \
--disable-multilib \
--enable-languages=c \
--disable-bootstrap \
--with-newlib

make -j $(nproc)

#do check 
make check-gcc RUNTESTFLAGS="rvp64.exp"
```

### CODE REVIEW

File: gcc/common/config/riscv/riscv-common.cc [hunk 1]
```diff
@@ -778,11 +785,16 @@ riscv_subset_list::add (const char *subset, int major_version,
...
+      if (std::strncmp("zpn", subset, 3)
+          && std::strncmp("zbpbo", subset, 5)
+          && std::strncmp("zpsfoperand", subset, 11))
+        {
+          error_at (m_loc,
+			"%<-march=%s%>: extension %qs starts with 'z' but is "
+			"unsupported standard extension",
+			m_arch, subset);
+          return;
+        }
```
Comment (yulong18) [inline gcc/common/config/riscv/riscv-common.cc]: 为什么要用两层if限制同一段代码呢？

File: gcc/config/riscv/riscv-builtins.cc [hunk 2]
```diff
@@ -129,6 +138,18 @@ AVAIL (zbb, TARGET_ZBB)
...
+AVAIL (zpn, TARGET_ZPN)
+AVAIL (zpn32, TARGET_ZPN && !TARGET_64BIT)
+AVAIL (zpn64, TARGET_ZPN && TARGET_64BIT)
+AVAIL (zbpbo, TARGET_ZBPBO)
+AVAIL (zbpbo32, TARGET_ZBPBO && !TARGET_64BIT)
+AVAIL (zbpbo64, TARGET_ZBPBO && TARGET_64BIT)
+AVAIL (zpsfoperand, TARGET_ZPSFOPERAND)
+AVAIL (zpsfoperand32, TARGET_ZPSFOPERAND && !TARGET_64BIT)
+AVAIL (zpsfoperand64, TARGET_ZPSFOPERAND && TARGET_64BIT)
+
```
Comment (yulong18) [inline gcc/config/riscv/riscv-builtins.cc:160]: 新增的内容放在别人代码的后面

File: gcc/config/riscv/riscv.cc [hunk 3]
```diff
@@ -8641,7 +8641,7 @@ riscv_sched_variable_issue (FILE *, int, rtx_insn *insn, int more)
 
   /* If we ever encounter an insn without an insn reservation, trip
      an assert so we can find and fix this problem.  */
-  gcc_assert (insn_has_dfa_reservation_p (insn));
+  //gcc_assert (insn_has_dfa_reservation_p (insn));
```
Comment (yulong18) [inline gcc/config/riscv/riscv.cc]: 不能随意注释掉别人的代码

**Review by yulong18 — COMMENTED** [review]
注意很多新增内容要放在别人原有的后面

File: gcc/config/riscv/riscv.cc (hunk 3 above)
Comment (zhixiao-zhang) [reply gcc/config/riscv/riscv.cc]: 我更改了实现方式，使用条件宏来控制这段代码

File: gcc/config/riscv/riscv-builtins.cc [hunk 4]
```diff
@@ -129,6 +138,7 @@ AVAIL (zbb, TARGET_ZBB)
 AVAIL (zbb64, TARGET_ZBB && TARGET_64BIT)
 AVAIL (zbb64_or_zbkb64, (TARGET_ZBKB || TARGET_ZBB) && TARGET_64BIT)
 AVAIL (zbb_or_zbkb, (TARGET_ZBKB || TARGET_ZBB))
+
```
Comment (yulong18) [inline gcc/config/riscv/riscv-builtins.cc:141]: 没特殊作用的话不修改原来的内容

**Review by yulong18 — COMMENTED** [review]
LGTM

**Pull request merged**